src/messenger/messenger.cpp

```diff
@@ -398,128 +404,225 @@ void Messenger::Terminate() {
+                                           GetBuiltinCallback(BuiltinCallback::ACK), identity_, "");
+        ZmqUtil::SendMsgIdentity(socket, identity_);
+        ZmqUtil::SendMsgPayload(socket, ack);
       }
-      // Otherwise, at least some socket has data. Is it the current socket?
-      bool socket_has_data = (item.revents & ZMQ_POLLIN) != 0;
-      if (socket_has_data) {
-        common::ManagedPointer<zmq::socket_t> socket(reinterpret_cast<zmq::socket_t *>(&item.socket));
-        ZmqMessage msg = ZmqUtil::RecvMsg(socket);
-        bool has_custom_serverloop = poll_items.server_callbacks_[i] != nullptr;
-        MESSENGER_LOG_TRACE("[PID={}] Messenger RECV-FR {} (custom serverloop: {}): {}", ::getpid(), msg.GetRoutingId(),
-                            has_custom_serverloop, msg.GetRawPayload());
-        // See the ProcessMessage docstring. It must always be invoked so that the callback that was passed in with
-        // SendMessage() is invoked.
-        ProcessMessage(msg);
+
+      bool has_custom_serverloop = poll_items.server_callbacks_[i] != nullptr;
+      MESSENGER_LOG_TRACE("[PID={}] Messenger RECV-FR {} (custom serverloop: {}): {}", ::getpid(), msg.GetRoutingId(),
+                          has_custom_serverloop, msg.GetRawPayload());
+      // See the ProcessMessage() docstring.
+      // ProcessMessage() must always be invoked so that the callback that was passed in with SendMessage() is invoked.
+      ProcessMessage(msg);
+      if (msg.GetDestinationCallbackId().UnderlyingValue() != static_cast<uint8_t>(BuiltinCallback::ACK)) {
+        std::string sender_id(msg.GetRoutingId());
+        bool first_time = UpdateMessagesSeen(sender_id, msg.GetMessageId());
+        if (!first_time) {
+          continue;
+        }
         if (has_custom_serverloop) {
           auto &server_callback = poll_items.server_callbacks_[i];
           (*server_callback)(common::ManagedPointer(this), msg);
         }
-        --num_sockets_with_data;
       }
+
+      --num_sockets_with_data;
+    }
+  }
+}
+
+bool Messenger::UpdateMessagesSeen(const std::string &replica, const message_id_t message_id) {
+  if (seen_messages_max_.find(replica) == seen_messages_max_.end()) {
+    seen_messages_max_.emplace(replica, message_id);
+    return true;
+  }
+
+  if (seen_messages_complement_.find(replica) == seen_messages_complement_.end()) {
+    seen_messages_complement_.emplace(replica, std::unordered_set<message_id_t>());
+  }
+
+  std::unordered_set<message_id_t> &complement = seen_messages_complement_.at(replica);
+  message_id_t &max_seen = seen_messages_max_.at(replica);
+
+  bool first_time = false;
+  if (message_id > max_seen) {
+    for (message_id_t i = max_seen + 1; i < message_id; ++i) {
+      complement.emplace(i);
+    }
+    max_seen = message_id;
+    first_time = true;
+  } else {
+    if (complement.find(message_id) != complement.end()) {
```

Comment:
Any particular reason that this isn't an `else if()`? Not really that important, but it could save you a couple curly braces.

Reply:
algorithm used to have an extra if, but that's not a thing now. refactored per suggestion, thanks!